Proposed product descriptions/Data_API_for_Genomes_product_description.md

```diff
@@ -0,0 +1,395 @@
+# KBase Data API for Genome data
+
+## Summary
+
+Currently any software written to use KBase data must be written
+explicitly against a set of data schemas using the KBase IDL schema
+language. The current data schemas are tailored to the specific needs of
+the services that initially created them, but are not a good general
+representation of the underlying concepts. Our current Workspace and IDL
+interfaces expose the service code from these implementation details,
+and make them "brittle" in the sense that any attempted change in the
+schemas will require manually coordinated changes in any system
+component that uses that data. This problem will be exacerbated by the
+potential wave of new "SDK" services from third party developers.
+
+The KBase Data API will address this weakness in our system by providing
+storage and schema-agnostic accessors for the data, which will enable
+the platform to evolve over time by providing a layer of abstraction
+over our current data stores and data schemas. High level API methods
+for accessing data for Genome related types eliminate the need to rely
+directly on specific data schemas for development. The API is versioned
+and documented to provide a clear contract with developers and users
+accessing data within KBase, and will address an endemic difficulty in
+the current platform with updating and comparing analyses through time.
+
+## Description
+
+A layer of abstraction is required to effectively hide direct data store
+and schema knowledge from the average developer/user to avoid code
+rewrites and redeployments for every component accessing data when
+changes inevitably occur. The KBase Data API will provide a single layer
+where data access occurs, relieving a significant burden on developers
+to track, test, and optimize for type changes in the system, and
+simplifying retrieval of that data through API methods. Reading the API
+documentation should be enough information for developers to perform
+tasks without needing to have intimate knowledge of how data is stored.
+
+## Requirements
+
+The basic requirements of the product are:
+
+-   Data type refactoring and normalization
+
+-   Unified set of accessors to both new (refactored) and old (legacy) types
+
+-   Capability to create/upload data for the new types
+
+-   Client/server deployment
+
+-   Support for languages: Python, JavaScript, Perl
```

Comment:
This is a representation of work done and essentially why, not the rainbow
of future possibilities.  Of course we would want to support Java as well.

On Mon, Mar 14, 2016 at 12:44 PM, MrCreosote <notifications@github.com>
wrote:

> In Proposed product
> descriptions/Data_API_for_Genomes_product_description.md
> <https://github.com/kbase/roadmap/pull/15#discussion_r56063054>:
>
> > +documentation should be enough information for developers to perform
> > +tasks without needing to have intimate knowledge of how data is stored.
> > +
> > +## Requirements
> > +
> > +The basic requirements of the product are:
> > +
> > +-   Data type refactoring and normalization
> > +
> > +-   Unified set of accessors to both new (refactored) and old (legacy) types
> > +
> > +-   Capability to create/upload data for the new types
> > +
> > +-   Client/server deployment
> > +
> > +-   Support for languages: Python, JavaScript, Perl
>
> No Java?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kbase/roadmap/pull/15/files#r56063054>.
>